enterprise/coderd/provisionerdaemons_test.go

```diff
@@ -794,4 +797,207 @@ func TestGetProvisionerDaemons(t *testing.T) {
+				tagsToFilterBy:        map[string]string{"scope": "user", "owner": "aaa", "environment": "on-prem"},
+				provisionerDaemonTags: map[string]string{"scope": "organization", "owner": ""},
+				expectToGetDaemon:     false,
+			},
+			{
+				name:                  "user-scoped provisioner and user-scoped filter with tags",
+				tagsToFilterBy:        map[string]string{"scope": "user", "owner": "aaa", "environment": "on-prem"},
+				provisionerDaemonTags: map[string]string{"scope": "user", "owner": "aaa"},
+				expectToGetDaemon:     false,
+			},
+			{
+				name:                  "user-scoped provisioner with tags and user-scoped filter with multiple tags",
+				tagsToFilterBy:        map[string]string{"scope": "user", "owner": "aaa", "environment": "on-prem", "datacenter": "chicago"},
+				provisionerDaemonTags: map[string]string{"scope": "user", "owner": "aaa", "environment": "on-prem"},
+				expectToGetDaemon:     false,
+			},
+			{
+				name:                  "user-scoped provisioner with tags and user-scoped filter with differing tags",
+				tagsToFilterBy:        map[string]string{"scope": "user", "owner": "aaa", "environment": "on-prem", "datacenter": "new_york"},
+				provisionerDaemonTags: map[string]string{"scope": "user", "owner": "aaa", "environment": "on-prem", "datacenter": "chicago"},
+				expectToGetDaemon:     false,
+			},
+		}
+		for _, tt := range testCases {
+			tt := tt
+			t.Run(tt.name, func(t *testing.T) {
+				t.Parallel()
+				dv := coderdtest.DeploymentValues(t)
+				client, db, _ := coderdenttest.NewWithDatabase(t, &coderdenttest.Options{
+					Options: &coderdtest.Options{
+						DeploymentValues: dv,
+					},
+					ProvisionerDaemonPSK: "provisionersftw",
+					LicenseOptions: &coderdenttest.LicenseOptions{
+						Features: license.Features{
+							codersdk.FeatureExternalProvisionerDaemons: 1,
+							codersdk.FeatureMultipleOrganizations:      1,
+						},
+					},
+				})
+				ctx := testutil.Context(t, testutil.WaitShort)
+
+				org := coderdenttest.CreateOrganization(t, client, coderdenttest.CreateOrganizationOptions{
+					IncludeProvisionerDaemon: false,
+				})
+				orgAdmin, _ := coderdtest.CreateAnotherUser(t, client, org.ID, rbac.ScopedRoleOrgMember(org.ID))
+
+				daemonCreatedAt := time.Now()
+
+				//nolint:gocritic // We're not testing auth on the following in this test
+				provisionerKey, err := db.InsertProvisionerKey(dbauthz.AsSystemRestricted(ctx), database.InsertProvisionerKeyParams{
+					Name:           "Test Provisioner Key",
+					ID:             uuid.New(),
+					CreatedAt:      daemonCreatedAt,
+					OrganizationID: org.ID,
+					HashedSecret:   []byte{},
+					Tags:           tt.provisionerDaemonTags,
+				})
+				require.NoError(t, err, "should be able to create a provisioner key")
+
+				//nolint:gocritic // We're not testing auth on the following in this test
+				pd, err := db.UpsertProvisionerDaemon(dbauthz.AsSystemRestricted(ctx), database.UpsertProvisionerDaemonParams{
```

Comment:
nit, non-blocking: I feel like you could stand up a single coderdtest deployment, create all of those provisioners in one go, and just test the filtering in subtests. Definitely not a blocker though!

Reply:
because we reuse tags in many of the test cases, some subsets of others, it would make the test assertions harder to deal with. Not an impossible problem, but probably not worth it unless I'm missing a benefit.

This does inspire me to test whether we deal with a group of differently tagged provisioners correctly though. Will add a test case.